File: rust/links-notation/src/stream_parser.rs
//! Incremental parsing for long-lived Links Notation streams.

use crate::parser::quoted_reference_end;
use crate::{parse_lino_to_links_with_config, LiNo, ParseError, ParserConfig};
use std::collections::VecDeque;
use std::fmt;

const DEFAULT_MAX_BUFFER_SIZE: usize = 10 * 1024 * 1024;
const BEFORE_REFERENCE: &[u8] = b" \t\n\r(:";
const BEFORE_COMMENT: &[u8] = b" \t\n\r";

/// The point immediately after the last character written to a stream parser.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct StreamPosition {
    /// One-based line number.
    pub line: usize,
    /// One-based character column.
    pub column: usize,
    /// Absolute byte offset from the start of the stream.
    pub offset: usize,
    /// Bytes held for the one unresolved top-level record.
    pub buffered: usize,
}

/// Backward-compatible name for stream position information.
pub type ErrorLocation = StreamPosition;

impl fmt::Display for StreamPosition {
    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(
            formatter,
            "line {}, column {} (offset {})",
            self.line, self.column, self.offset
        )
    }
}

/// A canonical parser or stream-limit error with absolute stream coordinates.
#[derive(Debug)]
pub struct StreamParseError {
    /// Human-readable description of the error.
    pub message: String,
    /// Position in the complete stream, when available.
    pub location: Option<StreamPosition>,
    /// The canonical parser error, when parsing a completed record failed.
    pub parse_error: Option<Box<ParseError>>,
}

impl fmt::Display for StreamParseError {
    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
        match &self.location {
            Some(location) => write!(
                formatter,
                "stream parse error at {location}: {}",
                self.message
            ),
            None => write!(formatter, "stream parse error: {}", self.message),
        }
    }
}

impl std::error::Error for StreamParseError {
    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
        self.parse_error
            .as_deref()
            .map(|error| error as &(dyn std::error::Error + 'static))
    }
}

impl From<ParseError> for StreamParseError {
    fn from(error: ParseError) -> Self {
        Self {
            message: error.to_string(),
            location: None,
            parse_error: Some(Box::new(error)),
        }
    }
}

/// Callback invoked for every completed link.
pub type LinkCallback = Box<dyn FnMut(&LiNo<String>) + Send>;

/// Callback invoked when a write or finish operation fails.
pub type ErrorCallback = Box<dyn FnMut(&StreamParseError) + Send>;

/// Incrementally emits complete top-level Links Notation records.
///
/// A record can acquire indented children, so a terminating newline is not by
/// itself a safe boundary. A record is committed when the next non-indented
/// content line begins, and the canonical parser validates every committed
/// segment. Set collection off for bounded-memory callback or iterator use.
pub struct StreamParser {
    config: ParserConfig,
    buffer: String,
    current_line: String,
    base_indentation: Option<usize>,
    line_classified: bool,
    link_callback: Option<LinkCallback>,
    error_callback: Option<ErrorCallback>,
    collect: bool,
    max_buffer_size: usize,
    links: Vec<LiNo<String>>,
    offset: usize,
    line: usize,
    column: usize,
    segment_offset: usize,
    segment_line: usize,
    ended: bool,
}

impl Default for StreamParser {
    fn default() -> Self {
        Self::new()
    }
}

impl StreamParser {
    /// Create a parser with the canonical default options.
    pub fn new() -> Self {
        Self::with_config(ParserConfig::default())
    }

    /// Create a parser with explicit canonical parser options.
    pub fn with_config(config: ParserConfig) -> Self {
        Self {
            config,
            buffer: String::new(),
            current_line: String::new(),
            base_indentation: None,
            line_classified: false,
            link_callback: None,
            error_callback: None,
            collect: true,
            max_buffer_size: DEFAULT_MAX_BUFFER_SIZE,
            links: Vec::new(),
            offset: 0,
            line: 1,
            column: 1,
            segment_offset: 0,
            segment_line: 1,
            ended: false,
        }
    }

    /// Register a callback invoked once for every completed link.
    pub fn on_link<F>(&mut self, callback: F) -> &mut Self
    where
        F: FnMut(&LiNo<String>) + Send + 'static,
    {
        self.link_callback = Some(Box::new(callback));
        self
    }

    /// Register a callback invoked when parsing fails.
    pub fn on_error<F>(&mut self, callback: F) -> &mut Self
    where
        F: FnMut(&StreamParseError) + Send + 'static,
    {
        self.error_callback = Some(Box::new(callback));
        self
    }

    /// Enable or disable retaining links for [`finish`](Self::finish) and [`drain`](Self::drain).
    pub fn set_collect(&mut self, collect: bool) -> &mut Self {
        self.collect = collect;
        self
    }

    /// Set the largest unresolved record accepted by the stream.
    pub fn set_max_buffer_size(
        &mut self,
        max_buffer_size: usize,
    ) -> Result<&mut Self, StreamParseError> {
        if max_buffer_size == 0 {
            return Err(StreamParseError {
                message: "maximum buffer size must be positive".to_string(),
                location: Some(self.position()),
                parse_error: None,
            });
        }
        self.max_buffer_size = max_buffer_size;
        Ok(self)
    }

    /// Consume a string chunk and return links made complete by it.
    pub fn write(&mut self, chunk: &str) -> Result<Vec<LiNo<String>>, StreamParseError> {
        if self.ended {
            return self.fail(StreamParseError {
                message: "cannot write after finish()".to_string(),
                location: Some(self.position()),
                parse_error: None,
            });
        }

        let mut emitted = Vec::new();
        for character in chunk.chars() {
            self.current_line.push(character);
            self.offset += character.len_utf8();

            if character == '\n' {
                self.buffer.push_str(&self.current_line);
                self.current_line.clear();
                self.line_classified = false;
                self.line += 1;
                self.column = 1;
            } else {
                if !self.line_classified && !matches!(character, ' ' | '\t' | '\r') {
                    self.line_classified = true;
                    if !(self.config.comments && character == '#') {
                        let indentation = leading_spaces(&self.current_line);
                        self.start_content_line(indentation, &mut emitted);
                    }
                }
                self.column += 1;
            }

            if self.buffer.len() + self.current_line.len() > self.max_buffer_size {
                return self.fail(StreamParseError {
                    message: format!(
                        "buffered record exceeds maximum size of {} bytes",
                        self.max_buffer_size
                    ),
                    location: Some(self.position()),
                    parse_error: None,
                });
            }
        }
        Ok(emitted)
    }

    /// Finish the stream and return all undrained or newly emitted links.
    pub fn finish(&mut self) -> Result<Vec<LiNo<String>>, StreamParseError> {
        if self.ended {
            return Ok(if self.collect {
                self.links.clone()
            } else {
                Vec::new()
            });
        }

        let mut emitted = Vec::new();
        let document = format!("{}{}", self.buffer, self.current_line);
        if !document.is_empty() {
            let parsed = match parse_lino_to_links_with_config(&document, &self.config) {
                Ok(links) => links,
                Err(error) => {
                    let stream_error = self.stream_error(error);
                    return self.fail(stream_error);
                }
            };
            self.publish(parsed, &mut emitted);
            self.advance_segment(&document);
        }

        self.buffer.clear();
        self.current_line.clear();
        self.base_indentation = None;
        self.ended = true;
        Ok(if self.collect {
            self.links.clone()
        } else {
            emitted
        })
    }

    /// Return and forget retained links.
    pub fn drain(&mut self) -> Vec<LiNo<String>> {
        std::mem::take(&mut self.links)
    }

    /// Reuse this parser while preserving its configuration and callbacks.
    pub fn reset(&mut self) -> &mut Self {
        self.buffer.clear();
        self.current_line.clear();
        self.base_indentation = None;
        self.line_classified = false;
        self.links.clear();
        self.offset = 0;
        self.line = 1;
        self.column = 1;
        self.segment_offset = 0;
        self.segment_line = 1;
        self.ended = false;
        self
    }

    /// Return the absolute stream position and unresolved buffer size.
    pub fn position(&self) -> StreamPosition {
        StreamPosition {
            line: self.line,
            column: self.column,
            offset: self.offset,
            buffered: self.buffer.len() + self.current_line.len(),
        }
    }

    /// Lazily parse any iterator of string-like chunks.
    pub fn parse_chunks<I, S>(chunks: I) -> StreamIterator<I::IntoIter>
    where
        I: IntoIterator<Item = S>,
        S: AsRef<str>,
    {
        let mut parser = Self::new();
        parser.set_collect(false);
        StreamIterator {
            chunks: chunks.into_iter(),
            parser,
            ready: VecDeque::new(),
            finished: false,
        }
    }

    fn start_content_line(&mut self, indentation: usize, emitted: &mut Vec<LiNo<String>>) {
        if !self.buffer.is_empty()
            && self
                .base_indentation
                .is_some_and(|base| indentation <= base)
            && structurally_complete(&self.buffer, self.config.comments)
        {
            if let Ok(parsed) = parse_lino_to_links_with_config(&self.buffer, &self.config) {
                let document = std::mem::take(&mut self.buffer);
                self.publish(parsed, emitted);
                self.advance_segment(&document);
                self.base_indentation = None;
            }
        }
        self.base_indentation.get_or_insert(indentation);
    }

    fn publish(&mut self, parsed: Vec<LiNo<String>>, emitted: &mut Vec<LiNo<String>>) {
        for link in parsed {
            emitted.push(link.clone());
            if self.collect {
                self.links.push(link.clone());
            }
            if let Some(callback) = &mut self.link_callback {
                callback(&link);
            }
        }
    }

    fn advance_segment(&mut self, document: &str) {
        self.segment_offset += document.len();
        self.segment_line += document.matches('\n').count();
    }

    fn stream_error(&self, error: ParseError) -> StreamParseError {
        let location = match &error {
            ParseError::SyntaxError(syntax) => StreamPosition {
                line: self.segment_line + syntax.line - 1,
                column: syntax.column,
                offset: self.segment_offset + syntax.offset,
                buffered: self.buffer.len() + self.current_line.len(),
            },
            _ => StreamPosition {
                line: self.segment_line,
                column: 1,
                offset: self.segment_offset,
                buffered: self.buffer.len() + self.current_line.len(),
            },
        };
        StreamParseError {
            message: error.to_string(),
            location: Some(location),
            parse_error: Some(Box::new(error)),
        }
    }

    fn fail<T>(&mut self, error: StreamParseError) -> Result<T, StreamParseError> {
        if let Some(callback) = &mut self.error_callback {
            callback(&error);
        }
        Err(error)
    }
}

/// Lazy iterator returned by [`StreamParser::parse_chunks`].
pub struct StreamIterator<I> {
    chunks: I,
    parser: StreamParser,
    ready: VecDeque<LiNo<String>>,
    finished: bool,
}

impl<I, S> Iterator for StreamIterator<I>
where
    I: Iterator<Item = S>,
    S: AsRef<str>,
{
    type Item = Result<LiNo<String>, StreamParseError>;

    fn next(&mut self) -> Option<Self::Item> {
        loop {
            if let Some(link) = self.ready.pop_front() {
                return Some(Ok(link));
            }
            if self.finished {
                return None;
            }
            match self.chunks.next() {
                Some(chunk) => match self.parser.write(chunk.as_ref()) {
                    Ok(links) => self.ready.extend(links),
                    Err(error) => {
                        self.finished = true;
                        return Some(Err(error));
                    }
                },
                None => {
                    self.finished = true;
                    match self.parser.finish() {
                        Ok(links) => self.ready.extend(links),
                        Err(error) => return Some(Err(error)),
                    }
                }
            }
        }
    }
}

fn leading_spaces(line: &str) -> usize {
    line.bytes().take_while(|byte| *byte == b' ').count()
}

fn follows(document: &[u8], position: usize, allowed: &[u8]) -> bool {
    position == 0 || allowed.contains(&document[position - 1])
}

fn structurally_complete(document: &str, comments: bool) -> bool {
    let bytes = document.as_bytes();
    let mut position = 0;
    let mut depth = 0_isize;

    while position < bytes.len() {
        let character = document[position..]
            .chars()
            .next()
            .expect("position is a character boundary");
        if matches!(character, '"' | '\'' | '`') && follows(bytes, position, BEFORE_REFERENCE) {
            let Some(end) = quoted_reference_end(document, position) else {
                return false;
            };
            position = end;
            continue;
        }
        if comments && character == '#' && follows(bytes, position, BEFORE_COMMENT) {
            match document[position..].find('\n') {
                Some(newline) => position += newline + 1,
                None => break,
            }
            continue;
        }
        if character == '(' {
            depth += 1;
        } else if character == ')' {
            depth -= 1;
        }
        position += character.len_utf8();
    }
    depth == 0
}

File: python/links_notation/stream_parser.py
"""Incremental parsing for long-lived Links Notation streams."""

from dataclasses import dataclass
from typing import AsyncIterable, AsyncIterator, Callable, Iterable, Iterator, List, Optional

from .link import Link
from .parser import DEFAULT_MAX_DEPTH, ParseError, Parser
from .quotes import QUOTE_CHARS, _parse_quoted_string_at


@dataclass(frozen=True)
class StreamPosition:
    """Position after the last character written to a stream parser."""

    offset: int
    line: int
    column: int
    buffered: int


class StreamParseError(ParseError):
    """A canonical parse error located within the complete stream.

    ``error`` is the error the parser raised for the buffered record;
    ``offset``, ``line`` and ``column`` locate it within the complete stream.
    """

    def __init__(self, error: Exception, offset: int, line: int, column: int):
        super().__init__(
            f"Stream parse error at line {line}, column {column}: {error}",
            offset=offset,
            line=line,
            column=column,
            line_text=getattr(error, "line_text", None),
            max_depth=getattr(error, "max_depth", None),
        )
        self.error = error


class StreamParser:
    """Incrementally emit complete top-level Links Notation records.

    A record can acquire indented children, so it becomes safe to emit only
    when the next non-indented content line begins. The regular :class:`Parser`
    remains the source of truth for each completed record.
    """

    def __init__(
        self,
        parser: Optional[Parser] = None,
        *,
        on_link: Optional[Callable[[Link], None]] = None,
        collect: bool = True,
        max_buffer_size: Optional[int] = None,
        comments: bool = True,
        max_depth: int = DEFAULT_MAX_DEPTH,
    ):
        self.parser = parser or Parser(comments=comments, max_depth=max_depth)
        self.on_link = on_link
        self.collect = collect
        self.max_buffer_size = self.parser.max_input_size if max_buffer_size is None else max_buffer_size
        self.reset()

    def write(self, chunk: str) -> List[Link]:
        """Consume a string chunk and return links completed by it."""
        if not isinstance(chunk, str):
            raise TypeError("Input must be a string")
        if self._ended:
            raise RuntimeError("Cannot write after finish()")

        emitted: List[Link] = []
        for character in chunk:
            self._current_line += character
            self._offset += 1

            if character == "\n":
                self._buffer += self._current_line
                self._current_line = ""
                self._line_classified = False
                self._line += 1
                self._column = 1
            else:
                if not self._line_classified and character not in (" ", "\t", "\r"):
                    self._line_classified = True
                    if not (self.parser.comments and character == "#"):
                        self._start_content_line(_leading_spaces(self._current_line), emitted)
                self._column += 1

            if len(self._buffer) + len(self._current_line) > self.max_buffer_size:
                raise ValueError(f"Buffered record exceeds maximum size of {self.max_buffer_size} characters")

        return emitted

    def finish(self, chunk: str = "") -> List[Link]:
        """Finish the stream and return all undrained or newly emitted links."""
        if self._ended:
            if chunk:
                raise RuntimeError("Cannot write after finish()")
            return list(self._links) if self.collect else []

        emitted = self.write(chunk) if chunk else []
        document = self._buffer + self._current_line
        if document:
            try:
                links = self.parser.parse(document)
            except Exception as error:
                raise self._stream_error(error) from error
            self._publish(links, emitted)
            self._advance_segment(document)

        self._buffer = ""
        self._current_line = ""
        self._base_indentation = None
        self._ended = True
        return list(self._links) if self.collect else emitted

    def drain(self) -> List[Link]:
        """Return and forget retained links."""
        links = self._links
        self._links = []
        return links

    def reset(self) -> "StreamParser":
        """Reuse the parser while preserving its configuration and callback."""
        self._buffer = ""
        self._current_line = ""
        self._base_indentation: Optional[int] = None
        self._line_classified = False
        self._links: List[Link] = []
        self._offset = 0
        self._line = 1
        self._column = 1
        self._segment_offset = 0
        self._segment_line = 1
        self._ended = False
        return self

    @property
    def position(self) -> StreamPosition:
        """Absolute position and current unresolved buffer size."""
        return StreamPosition(
            offset=self._offset,
            line=self._line,
            column=self._column,
            buffered=len(self._buffer) + len(self._current_line),
        )

    def _start_content_line(self, indentation: int, emitted: List[Link]) -> None:
        if (
            self._buffer
            and self._base_indentation is not None
            and indentation <= self._base_indentation
            and _structurally_complete(self._buffer, self.parser.comments)
        ):
            try:
                links = self.parser.parse(self._buffer)
            except Exception:
                links = None
            if links is not None:
                self._publish(links, emitted)
                self._advance_segment(self._buffer)
                self._buffer = ""
                self._base_indentation = None

        if self._base_indentation is None:
            self._base_indentation = indentation

    def _publish(self, links: List[Link], emitted: List[Link]) -> None:
        for link in links:
            emitted.append(link)
            if self.collect:
                self._links.append(link)
            if self.on_link is not None:
                self.on_link(link)

    def _stream_error(self, error: Exception) -> StreamParseError:
        """Locate an error the parser raised for the buffered record within the stream."""
        line = getattr(error, "line", None)
        column = getattr(error, "column", None)
        offset = getattr(error, "offset", None)
        if line is None or column is None or offset is None:
            return StreamParseError(error, self._segment_offset, self._segment_line, 1)
        return StreamParseError(error, self._segment_offset + offset, self._segment_line + line - 1, column)

    def _advance_segment(self, document: str) -> None:
        self._segment_offset += len(document)
        self._segment_line += document.count("\n")


def parse_chunks(chunks: Iterable[str], parser: Optional[Parser] = None) -> Iterator[Link]:
    """Lazily parse a synchronous iterable of chunks."""
    stream = StreamParser(parser, collect=False)
    for chunk in chunks:
        yield from stream.write(chunk)
    yield from stream.finish()


async def parse_async_chunks(chunks: AsyncIterable[str], parser: Optional[Parser] = None) -> AsyncIterator[Link]:
    """Lazily parse an asynchronous iterable of chunks."""
    stream = StreamParser(parser, collect=False)
    async for chunk in chunks:
        for link in stream.write(chunk):
            yield link
    for link in stream.finish():
        yield link


def _leading_spaces(line: str) -> int:
    return len(line) - len(line.lstrip(" "))


def _structurally_complete(document: str, comments: bool) -> bool:
    depth = 0
    position = 0
    before_reference = " \t\n\r(:"
    before_comment = " \t\n\r"

    while position < len(document):
        character = document[position]
        previous = None if position == 0 else document[position - 1]
        if character in QUOTE_CHARS and (previous is None or previous in before_reference):
            parsed = _parse_quoted_string_at(document, position)
            if parsed is None:
                return False
            position = parsed[1]
            continue
        if comments and character == "#" and (previous is None or previous in before_comment):
            newline = document.find("\n", position)
            if newline < 0:
                break
            position = newline + 1
            continue
        if character == "(":
            depth += 1
        elif character == ")":
            depth -= 1
        position += 1

    return depth == 0
